templates/templates.go

```diff
@@ -0,0 +1,418 @@
+	RoleTemplate         Template = "role.tmpl"
+	ResetEmailTemplate   Template = "resetEmail.tmpl" // generated by go generate
+	PermissionsTemplate  Template = "permissions.tmpl"
+	PermissionTemplate   Template = "permission.tmpl"
+
+	NoNavType      TemplateType = 0
+	PaginationType TemplateType = 1
+	RegularType    TemplateType = 2
+)
+
+func NewTemplate(p *permission.Store, r *role.Store, u *user.Store) *Templater {
+	return &Templater{
+		Permission: p,
+		Role:       r,
+		User:       u,
+	}
+}
+
+func (t Template) String() string {
+	return string(t)
+}
+
+func (t *Templater) RenderTemplate(w io.Writer, data interface{}, mainTmpl Template, templateType TemplateType) error {
+	var err error
+
+	t1 := template.New("_base.tmpl")
+
+	t1.Funcs(t.getFuncMaps())
+
+	switch templateType {
+	case NoNavType:
+		t1 = template.New("_baseNoNavs.tmpl")
+		t1.Funcs(t.getFuncMaps())
+		t1, err = t1.ParseFS(tmpls, "_baseNoNavs.tmpl", "_head.tmpl", "_footer.tmpl", string(mainTmpl))
+	case PaginationType:
+		t1, err = t1.ParseFS(tmpls, "_base.tmpl", "_head.tmpl", "_footer.tmpl", "_navbar.tmpl", "_sidebar.tmpl", "_pagination.tmpl", mainTmpl.String())
+	case RegularType:
+		t1, err = t1.ParseFS(tmpls, "_base.tmpl", "_head.tmpl", "_footer.tmpl", "_navbar.tmpl", "_sidebar.tmpl", mainTmpl.String())
+	default:
+		return fmt.Errorf("unable to parse template, invalid type: %d", templateType)
+	}
+
+	t1.Funcs(t.getFuncMaps())
+
+	if err != nil {
+		log.Printf("failed to get templates for template(RenderTemplate): %+v", err)
+		return err
+	}
+
+	return t1.Execute(w, data)
+}
+
+func (t *Templater) RenderEmail(emailTemplate Template) *template.Template {
+	return template.Must(template.New(emailTemplate.String()).ParseFS(tmpls, emailTemplate.String()))
+}
+
+func (t *Templater) getFuncMaps() template.FuncMap {
+	return template.FuncMap{
+		"formatDuration": func(d time.Duration) string {
+			days := int64(d.Hours()) / 24
+			hours := int64(d.Hours()) % 24
+			minutes := int64(d.Minutes()) % 60
+			seconds := int64(d.Seconds()) % 60
+
+			segments := []struct {
+				name  string
+				value int64
+			}{
+				{"Day", days},
+				{"Hour", hours},
+				{"Min", minutes},
+				{"Sec", seconds},
+			}
+
+			var parts []string
+
+			for _, s := range segments {
+				if s.value == 0 {
+					continue
+				}
+				plural := ""
+				if s.value != 1 {
+					plural = "s"
+				}
+
+				parts = append(parts, fmt.Sprintf("%d %s%s", s.value, s.name, plural))
+			}
+			return strings.Join(parts, " ")
+		},
+		"formatTime": func(fmt string, t time.Time) string {
+			return t.Format(fmt)
+		},
+		"now": func() time.Time {
+			return time.Now()
+		},
+		"thisYear": func() int {
+			return time.Now().Year()
+		},
+		"add": func(a, b int) int {
+			return a + b
+		},
+		"inc": func(a int) int {
+			return a + 1
+		},
+		"dec": func(a int) int {
+			return a - 1
+		},
+		"even": func(a int) bool {
+			return a%2 == 0
+		},
+		"len": func(a string) int {
+			return len(a)
+		},
```

Comment:
Do we need this if `text/template` has a `len` builtin?

Reply:
I didn't realise it was built in